gh-137578: support top-level setup statements in `Timer` objects

picnixz opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-137578 -->
* Issue: gh-137578
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--137587.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
support top-level setup statements in `Timer` objects

inventshah commented on Lib/timeit.py:
```diff
@@ -99,12 +106,27 @@ class Timer:
+            code = compile(global_setup, dummy_src_name, "exec")
+            exec(global_setup, global_ns, local_setup_ns)
+            # The global setup statements may have future statements.
+            compile_options['flags'] = code.co_flags
```
Since future flags can change what is compilable, should these options also be passed into the checks that `setup` and `stmt` can be compiled outside a function (L132/L144)?

Example `python -m timeit -g "from __future__ import barry_as_FLUFL" -- "1 <> 2"` currently fails.

picnixz replied:
Urgh. I think we're down to the same issue we had when we fixed barry_as_FLUFL on the REPL side. And yes they should be passed as well.

inventshah commented on Doc/library/timeit.rst:
```diff
@@ -62,30 +62,44 @@ Python Interface
-
-   Create a :class:`Timer` instance with the given statement, *setup* code and
-   *timer* function and run its :meth:`.timeit` method with *number* executions.
+.. function:: timeit(stmt='pass', setup='pass', timer=<default timer>,\
+                     number=1000000, globals=None,\
+                     *,\
+                     global_setup='pass')
```
Looks like the corresponding changes for `timeit` and `repeat` aren't in this PR, only updates to `Timer`?

picnixz replied:
Oh. I forgot.

inventshah commented on Doc/library/timeit.rst:
```diff
@@ -96,36 +110,52 @@ The module defines three convenience functions and a public class:
-   or newlines, as long as they don't contain multi-line string literals.  The
-   statement will by default be executed within timeit's namespace; this behavior
+
+   *stmt*, *setup* and *global_setup* may also contain multiple statements
+   separated by ``;`` or newlines, as long as they don't contain multi-line
+   string literals.
```
I know this has been in the docs for a while, but why aren't multi-line strings allowed? A quick test shows the implementation allows them, though I suppose strings inside `stmt` or `setup` may get indented differently than the user intends.

picnixz replied:
Because we reindent the user's code by replacing `\n` (it's a naive reindentation, not based on textwrap and possibly because of historical reasons). Maybe we should instead use `dedent()` followed by `indent()`, in which case that could be ok I guess. However, this should be a separate issue.

inventshah replied:
Hm, at least for the new `global_setup` that doesn't apply. But even for the existing parameters I'm not sure multi-line strings are prohibited as the docs suggest or just may produce unexpected results (due to reindenting). Regardless we can open a separate issue to discuss a docs change.

picnixz replied:
Yes, let's discuss it separately.

picnixz committed:
Update timeit.py

- carry co_flags obtained from global setup
- update timeit() and repeat() functions

picnixz reviewed (commented):
Ideally, I want to change the way we create the main() function because I think it'd be cleaner if we used argparse instead.

picnixz committed:
strip leading and trailing whitespaces from source

serhiy-storchaka commented on Lib/timeit.py:
```diff
@@ -99,15 +106,30 @@ class Timer:
+        if isinstance(global_setup, str):
+            # Check that the code can be compiled outside a function.
+            code = compile(global_setup, dummy_src_name, "exec")
+            exec(global_setup, global_ns, local_setup_ns)
```
Why execute it separately?

picnixz replied:
Are you talking about the code variable? or the local namespace?

picnixz replied:
Also it's separate because setup is executed at each repetition inside the inner() function so we need to make execute the global setup exactly once. And whatever becomes part of the local namespace for the global setup becomes part of the global namespace for setup and stmt.

picnixz replied:
Actually, it would have been fine to use the prefix approach as for the others, but the problem would be with variables declared at the global setup phase that aren't visible inside the `inner` function: `./python -m timeit -n1 -g 's=1' 'print(s)'` (they are not visible because we use a specific global dict and a specific local dict as well)

```py
s=1

def inner(_it, _timer):
    pass
    _t0 = _timer()
    for _i in _it:
        print(s)  # NameError
        pass
    _t1 = _timer()
    return _t1 - _t0
```

Therefore, I need to pick all variables that are declared in the global scope and make them available in the function (but `global_ns` doesn't have those variables *yet*). Or am I wrong here?

serhiy-storchaka replied:
The `inner` function will capture variables created by the global setup if use the same dict for globals and locals.
```py
        exec(code, global_ns, global_ns)
        self.inner = global_ns["inner"]
```
Of course, `inner` should be renamed to avoid possible conflicts.

I don't know what approach is better. I only say that there is other approach and we should consider it it see what is better.

picnixz replied:
I've considered renaming all the internal variables in the template with a suffix `os.urandom(8).hex()` but felt that it shouldn't have been part of this PR.

serhiy-storchaka replied:
Simply rename it to `_inner`. We already use `_it`, `_timer`, `_i`, `_t0`, `_t1`, `_setup`, `_stmt`.

serhiy-storchaka replied:
So what happens if we remove this `exec()`?

serhiy-storchaka reviewed (commented):
All LGTM, but we can try other approach before choosing the best.

It may be worth to mention that names created in the global setup will be seen as globals in the setup and testing code.

picnixz committed:
update docs

serhiy-storchaka commented on Doc/library/timeit.rst:
```diff
@@ -62,30 +62,46 @@ Python Interface
-   count and *number* executions.  The optional *globals* argument specifies a
-   namespace in which to execute the code.
+   Create a :class:`Timer` instance with the given statement, *setup* code,
+   *global_setup* global code and *timer* function and run its :meth:`.repeat`
```
`*global_setup* global code` looks weird.

Also, how would look in translation, when "global setup code" is translated?

Maybe remove emphasis around `*setup*` and leave "with the given statement and setup code"? We don't mention the *stmt* parameter by name here.

serhiy-storchaka commented on Lib/timeit.py:
```diff
@@ -99,15 +106,30 @@ class Timer:
+            global_setup = ''
+        elif callable(global_setup):
+            local_ns['_global_setup'] = global_setup
+            global_setup = '_global_setup(); del _global_setup'
```
We do not delete `_setup` after call.
